internal/noiseconn/conn.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package noiseconn contains an internal-only wrapper around controlbase.Conn
+// that properly handles the early payload sent by the server before the HTTP/2
+// session begins.
+//
+// See the documentation on the Conn type for more details.
+package noiseconn
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+	"sync"
+
+	"golang.org/x/net/http2"
+	"tailscale.com/control/controlbase"
+	"tailscale.com/tailcfg"
+)
+
+// Conn is a wrapper around controlbase.Conn.
+//
+// It allows attaching an ID to a connection to allow cleaning up references in
+// the pool when the connection is closed, properly handles an optional "early
+// payload" that's sent prior to beginning the HTTP/2 session, and provides a
+// way to return a connection to a pool when the connection is closed.
+type Conn struct {
+	*controlbase.Conn
+	id      int
+	onClose func(int)
+	h2cc    *http2.ClientConn
+
+	readHeaderOnce    sync.Once     // guards init of reader field
+	reader            io.Reader     // (effectively Conn.Reader after header)
+	earlyPayloadReady chan struct{} // closed after earlyPayload is set (including set to nil)
+	earlyPayload      *tailcfg.EarlyNoise
+	earlyPayloadErr   error
+}
+
+// New creates a new Conn that wraps the given controlbase.Conn.
+//
+// h2t is the HTTP/2 transport to use for the connection; a new
+// http2.ClientConn will be created that reads from the returned Conn.
+//
+// connID should be a unique ID for this connection. When the Conn is closed,
+// the onClose function will be called with the connID if it is non-nil.
+func New(conn *controlbase.Conn, h2t *http2.Transport, connID int, onClose func(int)) (*Conn, error) {
+	ncc := &Conn{
+		Conn:              conn,
+		id:                connID,
+		onClose:           onClose,
+		earlyPayloadReady: make(chan struct{}),
+	}
+	h2cc, err := h2t.NewClientConn(ncc)
+	if err != nil {
+		return nil, err
+	}
+	ncc.h2cc = h2cc
+	return ncc, nil
+}
+
+// RoundTrip implements the http.RoundTripper interface.
+func (c *Conn) RoundTrip(r *http.Request) (*http.Response, error) {
+	return c.h2cc.RoundTrip(r)
+}
+
+// GetEarlyPayload waits for the early noise payload to arrive.
+// It may return (nil, nil) if the server begins HTTP/2 without one.
+func (c *Conn) GetEarlyPayload(ctx context.Context) (*tailcfg.EarlyNoise, error) {
```

Comment:
say it's only valid to call this once at the beginning of the connection?

Reply:
Tested and confirmed; it's actually safe to call multiple times since the channel is closed to signal completion. Documented this 👍